secure/net/src/lib.rs

```diff
@@ -36,37 +51,38 @@ impl NetworkClient {
     }
 
     /// Blocking read until able to successfully read an entire message
-    pub fn read(&mut self) -> Result<Vec<u8>> {
+    pub fn read(&mut self) -> Result<Vec<u8>, Error> {
         let stream = self.server()?;
         let result = stream.read();
         if result.is_err() {
+            debug!("On read, upstream peer disconnected, setting stream to None");
             self.stream = None;
         }
         result
     }
 
     /// Shutdown the internal network stream
-    pub fn shutdown(&mut self) -> Result<()> {
-        let stream = self
-            .stream
-            .take()
-            .ok_or_else(|| anyhow!("No active stream"))?;
+    pub fn shutdown(&mut self) -> Result<(), Error> {
+        debug!("Shutdown called");
+        let stream = self.stream.take().ok_or_else(|| Error::NoActiveStream)?;
         stream.shutdown()?;
         Ok(())
     }
 
     /// Blocking read until able to successfully read an entire message
-    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+    pub fn write(&mut self, data: &[u8]) -> Result<(), Error> {
         let stream = self.server()?;
         let result = stream.write(data);
         if result.is_err() {
+            debug!("On write, upstream peer disconnected, setting stream to None");
             self.stream = None;
         }
         result
     }
 
-    fn server(&mut self) -> Result<&mut NetworkStream> {
+    fn server(&mut self) -> Result<&mut NetworkStream, Error> {
         if self.stream.is_none() {
+            debug!("Attempting to connect to upstream {}", self.server);
```

Comment:
log sth inside the while loop? (that's where it failed

Reply:
oh, hmm, I hate that kinda stuff lol, but we can add it for tracing level.